src/preprocessing/passes/daneshvar.cpp

```diff
@@ -0,0 +1,663 @@
+
+
+namespace cvc5::internal {
+namespace preprocessing {
+namespace passes {
+
+Daneshvar::Daneshvar(PreprocessingPassContext* preprocContext)
+    : PreprocessingPass(preprocContext, "daneshvar"),
+    d_statistics(statisticsRegistry())
+    {};
+
+
+
+
+// void dfs(
+//     const Node&n, 
+//     std::map<Node, uint32_t>& subtreeCache, // key: node, value: subtree id
+//     std::map<std::string, uint32_t>& symbolMap, // key: symbol, value: id (color)
+//     std::map<uint32_t, std::vector<std::string>>& subtreePattern, // key: subtree id, value: pattern (each element is a symbol color or subtree id)
+//     std::map<uint32_t, std::vector<std::string>>& symbolList // key: subtree id, value: list of symbols in the subtree
+//     )
+// {
+//     if (n.isVar())
+//     {
+//         std::string symbol = n.toString();
+//         if (symbolMap.find(symbol) == symbolMap.end())
+//         {
+//             uint32_t id = symbolMap.size() + 1;
+//             symbolMap[symbol] = id;
+//         }
+        
+//         return;
+//     }
+
+
+//     cvc5::internal::Kind k = static_cast<cvc5::internal::Kind>(n.getKind());
+//     std::string boz = cvc5::internal::kind::toString(k);
+//     std::vector<std::string> children;
+//     std::vector<std::string> symbols;
+//     for (size_t i = 0; i < n.getNumChildren(); i++)
+//     {
+//         if (n[i].isConst())
+//         {
+//             continue;
+//         }
+
+        
+//         if (n[i].isVar())
+//         {
+//             if (symbolMap.find(n[i].toString()) == symbolMap.end())
+//             {
+//                 dfs(n[i], subtreeCache, symbolMap, subtreePattern, symbolList);
+//             }
+
+//             children.push_back(std::to_string(symbolMap[n[i].toString()]));
+//             symbols.push_back(n[i].toString());
+//             continue;
+//         }
+        
+
+//         if (subtreeCache.find(n[i]) == subtreeCache.end())
+//         {
+//             dfs(n[i], subtreeCache, symbolMap, subtreePattern, symbolList);
+//         } 
+//         children.push_back("S" + std::to_string(subtreeCache[n[i]]));
+//         for (auto& elem : symbolList[subtreeCache[n[i]]])
+//         {
+//             symbols.push_back(elem);
+//         }
+        
+//     }
+    
+//     uint32_t id = subtreeCache.size() + 1;
+//     subtreeCache[n] = id;
+//     // std::cout << "Inserting " << n << " with id " << id << std::endl;
+//     subtreePattern[id] = children;
+//     symbolList[id] = symbols;
+// }
+
+
+
+
+void dfs_iterative(
+    const Node& root,
+    std::map<Node, uint32_t>& subtreeCache,
+    std::map<std::string, uint32_t>& symbolMap,
+    std::map<uint32_t, std::vector<std::string>>& subtreePattern,
+    std::map<uint32_t, std::vector<std::string>>& symbolList)
+{
+    // Stack to hold nodes and a flag indicating if the node's children have been processed
+    std::stack<std::pair<Node, bool>> stack;
+    stack.push({root, false});
+
+    while (!stack.empty())
```

Comment:
This also can to a tree-traversal, when going down. You need a map to cache the current state of the node (whether it was visited once or twice) instead of a boolean flag on the stack.

A common pattern for this is:


```
std::vector<Node> visit;
std::unordered_map<Node, bool> cache;

visit.push_back(start_node);
while (!visit.empty())
{
  TNode cur = visit.back();

  auto [it, inserted] = cache.emplace_back(cur, false);
  if (inserted) // traverse down
  {
    // process cur when traversing down
    visit.insert(visit.end(), cur.begin(), cur.end());
    continue;
  }
  else if (!it->second) // traverse up
  {
    it->second = true;
    // process cur when traversing up
  }
  visit.pop_back();
}
```

In your case above, you can probably use subtreeCache for that if id = 0 is not used, which is then used to indicate the state where a node is visited once when traversing down.